### PR TITLE
Feature/bar chart

### PR DESCRIPTION
**Pull request by mroncalloCondor**
Added EmeraldChart component

### CODE REVIEW

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    internal var colorsDictionary: [String: UIColor]!
+    internal var dataEntries: [EmeraldChartSimpleDataEntry]?
+    internal var subtitlesFont: UIFont?
+    
+    init(viewPortHandler: ViewPortHandler, xAxis: XAxis?, transformer: Transformer?, icon: UIImage) {
+        self.icon = icon
+        super.init(viewPortHandler: viewPortHandler, xAxis: xAxis, transformer: transformer)
+    }
+    
+    override func drawLabel(context: CGContext, formattedLabel: String, x: CGFloat, y: CGFloat, attributes: [NSAttributedString.Key : Any], constrainedToSize: CGSize, anchor: CGPoint, angleRadians: CGFloat) {
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Split into multiple lines

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        
+        super.drawLabel(context: context, formattedLabel: formattedLabel, x: x, y: y, attributes: attributes, constrainedToSize: constrainedToSize, anchor: anchor, angleRadians: angleRadians)
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: same here

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 1 above)
Comment (gsanguino) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: This is a library method or something? if not please rename params to a more descriptive name pls

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        let labelSize = (formattedLabel as NSString).boundingRect(with: constrainedToSize, options: .usesLineFragmentOrigin, attributes: attributes, context: nil)
+        
+        var subtitle = ""
+        
+        var entry: EmeraldChartSimpleDataEntry? = nil
+        if let entries = dataEntries {
+            entry = entries.first(where: {$0.title == formattedLabel})
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: What if I have more than one entry that has the same title?

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        }
+        
+        var subtitlesAttributes = attributes
+        if let foundEntry = entry {
+            if let subtitleFont = subtitlesFont {
+                subtitlesAttributes[NSAttributedString.Key.font] =  subtitleFont
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Why your validate param is singular and the optional one is plural?

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            }
+            subtitlesAttributes[NSAttributedString.Key.foregroundColor] = foundEntry.color
+            subtitle = String(Int(foundEntry.value))
+        }
+       
+        
+        if subtitle != "" {
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: if this is a string please use a .isEmpty instead..

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            super.drawLabel(context: context, formattedLabel: subtitle, x: x + 5, y: y + 20, attributes: subtitlesAttributes, constrainedToSize: constrainedToSize, anchor: anchor, angleRadians: angleRadians)
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: So you call twice the super??? why?

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        }
+        
+        if icon != UIImage() {
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: ??? you should validate if it is nil instead

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            ChartUtils.drawImage(context: context, image: icon, x: x + labelSize.width/1.5, y: y + 6, size: CGSize(width: 10, height: 14))
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Split into multiple lines and put all the Int into more descriptive Inner constants

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+
+import Foundation
+import Charts
+
+class EmeraldCustomBarChartRenderer: BarChartRenderer {
+    internal var cornerRadius: CGFloat = 0
+    private class Buffer {
+        var rects = [CGRect]()
+    }
+    private var _buffers = [Buffer]()
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: why you call the variables with initial underscore? this is not part of our guidelines..

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift [hunk 10]
```diff
@@ -0,0 +1,611 @@
...
+    private var _barShadowRectBuffer: CGRect = CGRect()
+    
+    override func initBuffers() {
+        if let barData = dataProvider?.barData
+        {
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: please all the brackets on this file should be right on side with the method, closure, etc

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift [hunk 11]
```diff
@@ -0,0 +1,611 @@
...
+                 * |             |  |  |   |//|    |                               |             |  +--+   |//|    |                               |
+                 * |             |  +--+   |//|    |                               |             |  |  |   |//|    |                               |
+                 * |             |         |//|    |                               |             |  |  |   |//|    |                               |
+                 * |      min -> +---------+--+----+   <- bottom + bottomOffset    |      max -> +--+--+---+--+----+    <- bottom - bottomOffset   |
+                 * |                       |//|         } bottomOffset = min - y   |                |  |   |//|          } bottomOffset = -max     |
+                 * |        y ->           +--+        <- bottom                   |        0 -> ---+--+---+--+-------  <- bottom                  |
+                 * |                                                               |                                                               |
+                 * +---------------------------------------------------------------+---------------------------------------------------------------+
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: remove comments pls

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+public struct EmeraldChartOptions {
+    public var cornerRadius: CGFloat
+    public var titleFont: UIFont
+    public var subtitleFont: UIFont
+    public var accessoryViewImage: UIImage
+    public var showSubtitle: Bool
+    public var showGrid: Bool
+    public var showLeftAxis: Bool
+    public var steps: Int
+    public var drawBorder: Bool
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift]: all of these need to be mutable?

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    
+    public init (cornerRadius: CGFloat = 0,
+                 titleFont: UIFont = UIFont.systemFont(ofSize: 10),
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift]: replace default by emerald fonts pls

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+                 subtitleFont: UIFont = UIFont.boldSystemFont(ofSize: 20),
+                 accessoryViewImage: UIImage = UIImage(),
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift]: this default value should be nil

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+                 showSubtitle: Bool = false,
+                 showGrid: Bool = false,
+                 showLeftAxis: Bool = true,
+                 steps: Int = 5,
+                 drawBorder: Bool = false){
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift]: space between method and braket

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+import Foundation
+
+class EmeraldChartPresenter: EmeraldChartPresenterType {
+    
+    private var chartView: EmeraldChart!
+    private var simpleDataEntries: [EmeraldChartSimpleDataEntry]?
+    private var multipleValueDataEntries: [EmeraldChartMultipleValueDataEntry]?
+    private var multipleValueColors = [EmeraldTheme.Chart.correctData, EmeraldTheme.Chart.warningData, EmeraldTheme.Chart.criticalData]
+    
+    init() {
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift]: you need this?

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+        }
+        return colors
+    }
+    
+    func getValueForMultipleValueDataEntry(index: Int) -> [Double] {
+        if let dataEntries = multipleValueDataEntries {
+            let values = dataEntries[index].value.compactMap{value in Double(value)}
+            return values
+        }
+        return [Double]()
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift]: this should return nil so we dont need to validate if its value are empty right?

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartView.swift [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+
+import UIKit
+import Charts
+
+public class EmeraldChart: UIView {
+    private var presenter: EmeraldChartPresenterType!
+    private var barChartView: BarChartView!
+    internal var delegate: EmeraldChartViewDelegate?
+    private var chartOptions: EmeraldChartOptions
+    var barChartRenderer: EmeraldCustomBarChartRenderer!
```
Comment (gsanguino) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartView.swift]: all needs to be var?

File: EmeraldIOS/Media.xcassets/Images/Edit.imageset/Contents.json [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+      "scale" : "1x"
+    },
+    {
+      "idiom" : "universal",
+      "filename" : "Edit@2x.png",
+      "scale" : "2x"
+    },
+    {
+      "idiom" : "universal",
+      "filename" : "Edit@3x.png",
```
Comment (gsanguino) [inline EmeraldIOS/Media.xcassets/Images/Edit.imageset/Contents.json]: what is this icon? what edit means? please use something like ---> descriptivename_icon

File: EmeraldIOS/Theme/Theme.swift [hunk 20]
```diff
@@ -73,6 +73,12 @@ public class EmeraldTheme {
         public static let downyColor = #colorLiteral(red: 0.4431372549, green: 0.8352941176, blue: 0.7490196078, alpha: 1)
     }
     
+    public struct Chart {
+        public static let criticalData = UIColor(red: 221/255, green: 53/255, blue: 10/255, alpha: 1)
```
Comment (gsanguino) [inline EmeraldIOS/Theme/Theme.swift]: please use emerald colors instead (green,yellow, red)

File: EmeraldIOS/Utils/Extensions+UIColorDarkener.swift [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+//  EmeraldIOS
+//
+//  Created by Miguel Roncallo on 9/26/19.
+//  Copyright © 2019 Condor Labs. All rights reserved.
+//
+
+import Foundation
+
+extension UIColor {
+
```
Comment (gsanguino) [inline EmeraldIOS/Utils/Extensions+UIColorDarkener.swift]: please indent entire file

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 1 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: This is an override of the library's method

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 6 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: The library's "super" method has some internal protection code required to draw labels. The second call is used to draw the subtitle text

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift (hunk 9 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: This is a copy of the library's class with a small tweak to paint rounded corners bars

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift (hunk 12 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift]: Yes, this exposes all the modification available for the chart component

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift [hunk 22]
```diff
@@ -0,0 +1,611 @@
...
+                {
+                    _buffers.append(Buffer())
+                }
+                while _buffers.count > barData.dataSetCount
+                {
+                    _buffers.removeLast()
+                }
+            }
+            
+            for i in stride(from: 0, to: barData.dataSetCount, by: 1)
```
Comment (lgoyescondor) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: Taking into account that the stride increasing rate is `1`, don't you like the most using:
```suggestion
            for i in 0..<barData.dataSetCount {
```

Besides! you could use something shorter:
```suggestion
            for set in barData.dataSets where set is IBarChartDataSet {
```

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift [hunk 23]
```diff
@@ -0,0 +1,611 @@
...
+        let containsStacks = dataSet.isStacked
+        
+        let isInverted = dataProvider.isInverted(axis: dataSet.axisDependency)
+        let phaseY = animator.phaseY
+        var barRect = CGRect()
+        var x: Double
+        var y: Double
+
+        
+        for i in stride(from: 0, to: min(Int(ceil(Double(dataSet.entryCount) * animator.phaseX)), dataSet.entryCount), by: 1)
```
Comment (lgoyescondor) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: This is a good place to use the stride method

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift [hunk 24]
```diff
@@ -0,0 +1,611 @@
...
+        {
+            guard let e = dataSet.entryForIndex(i) as? BarChartDataEntry else { continue }
```
Comment (lgoyescondor) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: `e` is not a descriptive name

File: ProofOfConcept/ProofOfConcept/BarChartViewController.swift [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Do any additional setup after loading the view.
+        setupSimpleBarChartView()
+        setupMultipleValuesBarChartView()
+    }
+    
+    private func setupSimpleBarChartView() {
+        var dataEntries = [EmeraldChartSimpleDataEntry]()
+        dataEntries.append(EmeraldChartSimpleDataEntry(color: EmeraldTheme.blueColor , value: 20, title: "First Data", subtitle: nil, hasAction: false))
```
Comment (lgoyescondor) [inline ProofOfConcept/ProofOfConcept/BarChartViewController.swift]: Break lines. Try not to make them longer than 120 char

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+//  Created by Miguel Roncallo on 9/20/19.
+//  Copyright © 2019 Condor Labs. All rights reserved.
+//
+
+import Foundation
+
+class EmeraldChartPresenter: EmeraldChartPresenterType {
+    
+    private var chartView: EmeraldChart!
+    private var simpleDataEntries: [EmeraldChartSimpleDataEntry]?
```
Comment (lgoyescondor) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift]: I would prefer to have a single class `EmeraldChartDataEntry` that had inside another two variables `SingleEntry` and `MultipleEntry`.

The presenter shoult not handle the different between those types of entries. The `EmeraldChartDataEntry` should be responsible of handling that.

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+        return 0
+    }
+    
+    func bind(view: EmeraldChart) {
+        self.chartView = view
+    }
+    
+    func getDataHorizontalEntries() -> [String] {
+        var labels = [String]()
+        if let dataEntries = simpleDataEntries {
```
Comment (lgoyescondor) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift]: This should be handled by the `DataEntry` model

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartView.swift [hunk 28]
```diff
@@ -0,0 +1,120 @@
...
+    internal var delegate: EmeraldChartViewDelegate?
+    private let chartOptions: EmeraldChartOptions
+    var barChartRenderer: EmeraldCustomBarChartRenderer!
+    
+    public init(simpleData: [EmeraldChartSimpleDataEntry],
+                withOptions chartOptions: EmeraldChartOptions = EmeraldChartOptions()) {
+        self.chartOptions = chartOptions
+        super.init(frame: .zero)
+        self.presenter = EmeraldChartPresenter()
+        self.presenter.bind(view: self)
```
Comment (lgoyescondor) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartView.swift]: I'm not sure if this binding should take place in this moment.

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartView.swift [hunk 29]
```diff
@@ -0,0 +1,120 @@
...
+    
+    private func configBarChartView() {
+        setChartOptions()
+        barChartView.delegate = self
+        barChartView.maxVisibleCount = presenter.getDataEntriesCount()
+        barChartView.animate(yAxisDuration: 1)
+        barChartView.xAxis.gridColor = .clear
+        barChartView.xAxis.labelPosition = .bottom
+        barChartView.xAxis.labelHeight = 100
+        barChartView.extraBottomOffset = 40
```
Comment (lgoyescondor) [inline EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartView.swift]: Make all these things constants

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift [hunk 30]
```diff
@@ -0,0 +1,513 @@
...
+                        context.setFillColor(dataSet.barShadowColor.cgColor)
+                        context.fill(_barShadowRectBuffer)
+                    }
+                }
+
+                let buffer = _buffers[index]
+                
+                // draw the bar shadow before the values
+                if dataProvider.isDrawBarShadowEnabled {
+                    for j in stride(from: 0, to: buffer.rects.count, by: 1)  {
```
Comment (lgoyescondor) [inline EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: The stride is not necessary

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift [hunk 31]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    
+    override func drawLabel(context: CGContext, formattedLabel: String, x: CGFloat, y: CGFloat, attributes: [NSAttributedString.Key : Any], constrainedToSize: CGSize, anchor: CGPoint, angleRadians: CGFloat) {
+        
+        super.drawLabel(context: context, formattedLabel: formattedLabel, x: x, y: y, attributes: attributes, constrainedToSize: constrainedToSize, anchor: anchor, angleRadians: angleRadians)
+        let labelSize = (formattedLabel as NSString).boundingRect(with: constrainedToSize, options: .usesLineFragmentOrigin, attributes: attributes, context: nil)
+        
+        var subtitle = ""
+        
+        var entry: EmeraldChartSimpleDataEntry? = nil
```
[the remark this answers is not in the document]
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: This entry is only for the subtitles. It still paints the title label

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift [hunk 32]
```diff
@@ -0,0 +1,54 @@
...
+import Foundation
+import Charts
+
+class EmeraldChartXAxisRenderer: XAxisRenderer {
+    internal var icon: UIImage
+    internal var colorsDictionary: [String: UIColor]!
+    internal var dataEntries: [EmeraldChartSimpleDataEntry]?
+    internal var subtitlesFont: UIFont?
+    
+    init(viewPortHandler: ViewPortHandler, xAxis: XAxis?, transformer: Transformer?, icon: UIImage) {
```
[the remark this answers is not in the document]
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 1 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 2 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 4 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 5 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 6 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 7 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 8 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift (hunk 9 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift (hunk 10 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift (hunk 11 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift (hunk 13 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift]: Done

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift (hunk 14 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift]: Done

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift (hunk 15 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartOptions.swift]: Done

File: EmeraldIOS/Media.xcassets/Images/Edit.imageset/Contents.json (hunk 19 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Media.xcassets/Images/Edit.imageset/Contents.json]: Done

File: EmeraldIOS/Theme/Theme.swift (hunk 20 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Theme/Theme.swift]: Done

File: EmeraldIOS/Utils/Extensions+UIColorDarkener.swift (hunk 21 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Utils/Extensions+UIColorDarkener.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 31 above)
[the remark this answers is not in the document]
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift (hunk 32 above)
[the remark this answers is not in the document]
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldChartXAxISRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift [hunk 33]
```diff
@@ -0,0 +1,611 @@
...
+        var rects = [CGRect]()
+    }
+    private var _buffers = [Buffer]()
+    private var _barShadowRectBuffer: CGRect = CGRect()
+    
+    override func initBuffers() {
+        if let barData = dataProvider?.barData
+        {
+            // Matche buffers count to dataset count
+            if _buffers.count != barData.dataSetCount
```
[the remark this answers is not in the document]
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift (hunk 22 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift (hunk 24 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: Done

File: ProofOfConcept/ProofOfConcept/BarChartViewController.swift (hunk 25 above)
Comment (mroncalloCondor) [reply ProofOfConcept/ProofOfConcept/BarChartViewController.swift]: Done

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift (hunk 26 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift]: Done

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift (hunk 27 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift]: Done

File: EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift (hunk 30 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/CustomRenderers/EmeraldCustomBarChartRenderer.swift]: Done

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartView.swift (hunk 29 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartView.swift]: Done

File: EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift (hunk 16 above)
Comment (mroncalloCondor) [reply EmeraldIOS/Components/Chart/EmeraldChartCreator/EmeraldChartPresenter.swift]: Done

**Pull request merged**